libs/langgraph/langgraph/graph/state.py

```diff
@@ -830,3 +819,23 @@ def _get_schema(
                     if k in channels and isinstance(channels[k], BaseChannel)
                 },
             )
+
+
+def _get_input_schema_from_type_hint(
+    action: Optional[RunnableLike],
+) -> Optional[Type[Any]]:
+    try:
+        if not hasattr(action, "__call__"):
+            raise AttributeError
```

Comment:
could you populate these with descriptive error messages?

Reply:
Hi @vbarda, You're right.
I adjusted the code into the following two versions.
Is it better than the original? Which style do you prefer?
Both have the same logic, with only slight differences.

```py
def _get_input_schema_from_type_hint(
    action: Optional[RunnableLike],
) -> Optional[Type[Any]]:
    if not isfunction(action) and not ismethod(getattr(action, "__call__", None)):
        return None
    action = cast(Callable, action)

    hints = get_type_hints(getattr(action, "__call__")) or get_type_hints(action)
    try:
        first_parameter_name = next(iter(inspect.signature(action).parameters.keys()))
    except StopIteration:
        return None
    else:
        input_hint = hints.get(first_parameter_name)
        if isinstance(input_hint, type) and get_type_hints(input_hint):
            return input_hint
    return None
```

---

```py
def _get_input_schema_from_type_hint(
    action: Optional[RunnableLike],
) -> Optional[Type[Any]]:
    if not isfunction(action) and not ismethod(getattr(action, "__call__", None)):
        return None
    action = cast(Callable, action)

    hints = get_type_hints(getattr(action, "__call__")) or get_type_hints(action)
    try:
        first_parameter_name = next(iter(inspect.signature(action).parameters.keys()))
        input_hint = hints.get(first_parameter_name)
        if isinstance(input_hint, type) and get_type_hints(input_hint):
            return input_hint
    except StopIteration:
        pass        
    return None
```

Reply:
thank you -- personal preference for the second one :)

Reply:
Updated!